lib/stats/stats-cluster-key-builder.c

```diff
@@ -0,0 +1,182 @@
+
+void
+stats_cluster_key_builder_set_name_suffix(StatsClusterKeyBuilder *self, const gchar *name_suffix)
+{
+  g_free(self->name_suffix);
+  self->name_suffix = g_strdup(name_suffix);
+}
+
+void
+stats_cluster_key_builder_add_label(StatsClusterKeyBuilder *self, const StatsClusterLabel label)
+{
+  StatsClusterLabel own_label = stats_cluster_label(g_strdup(label.name), g_strdup(label.value));
+  self->labels = g_array_append_vals(self->labels, &own_label, 1);
+}
+
+void
+stats_cluster_key_builder_set_unit(StatsClusterKeyBuilder *self, StatsClusterUnit unit)
+{
+  self->unit = unit;
+}
+
+void
+stats_cluster_key_builder_set_legacy_alias(StatsClusterKeyBuilder *self, guint16 component, const gchar *id,
+                                           const gchar *instance)
+{
+  self->legacy.set = TRUE;
+  self->legacy.component = component;
+
+  g_free(self->legacy.id);
+  self->legacy.id = g_strdup(id);
+
+  g_free(self->legacy.instance);
+  self->legacy.instance = g_strdup(instance);
+}
+
+void
+stats_cluster_key_builder_set_legacy_alias_name(StatsClusterKeyBuilder *self, const gchar *name)
+{
+  self->legacy.set = TRUE;
+
+  g_free(self->legacy.name);
+  self->legacy.name = g_strdup(name);
+}
+
+static gint
+_labels_sort(const StatsClusterLabel *a, const StatsClusterLabel *b)
+{
+  return strcmp(a->name, b->name);
+}
+
+static gchar *
+_format_name(StatsClusterKeyBuilder *self)
+{
+  if (self->name_suffix)
+    return g_strdup_printf("%s%s", self->name, self->name_suffix);
+
+  return g_strdup(self->name);
+}
+
+void
+stats_cluster_key_builder_build_single(StatsClusterKeyBuilder *self, StatsClusterKey *sc_key)
+{
+  gchar *name = _format_name(self);
+  g_array_sort(self->labels, (GCompareFunc) _labels_sort);
+
+  StatsClusterKey temp_key;
+  stats_cluster_single_key_set(&temp_key, name, (StatsClusterLabel *) self->labels->data, self->labels->len);
+  if (self->legacy.set)
+    {
+      if (self->legacy.name)
+        stats_cluster_single_key_add_legacy_alias_with_name(&temp_key, self->legacy.component, self->legacy.id,
+                                                            self->legacy.instance, self->legacy.name);
+      else
+        stats_cluster_single_key_add_legacy_alias(&temp_key, self->legacy.component, self->legacy.id,
+                                                  self->legacy.instance);
+    }
+  stats_cluster_single_key_add_unit(&temp_key, self->unit);
+  stats_cluster_key_clone(sc_key, &temp_key);
+
+  g_free(name);
+  return;
+}
+
+void
+stats_cluster_key_builder_build_logpipe(StatsClusterKeyBuilder *self, StatsClusterKey *sc_key)
+{
+  gchar *name = _format_name(self);
+  g_array_sort(self->labels, (GCompareFunc) _labels_sort);
+
+  StatsClusterKey temp_key;
+  stats_cluster_logpipe_key_set(&temp_key, name, (StatsClusterLabel *) self->labels->data, self->labels->len);
+  if (self->legacy.set)
+    {
+      stats_cluster_logpipe_key_add_legacy_alias(&temp_key, self->legacy.component, self->legacy.id, self->legacy.instance);
```

Comment:
Should we add a `!self->legacy.name` assert?

Reply:
I was checking why the unit test passes like this, and had a mini heart attack, because the legacy name is not checked in `stats_cluster_key_equal()`, and on the first glance, I think it should be. I need a bit more time to think this through, let's discuss it tomorrow :)

Reply:
The name is for single keys, group_equals use this correctly.
I think we should just add an assert for the logpipe builder.